packages/@okta/vuepress-site/docs/guides/oie-embedded-widget-use-cases/oie-embedded-widget-use-case-sign-in-soc-idp/java/integrationsteps.md

```diff
@@ -0,0 +1,82 @@
+
+### Step 5: Handle the callback from Okta
+
+Okta returns the interaction code to the **Sign-in redirect URI** specified in the [create new application](/docs/guides/oie-embedded-common-org-setup/java/main/#step-4-create-new-application) step.
+
+```java
+String issuer = oktaOAuth2Properties.getIssuer();
+// the widget needs the base url, just grab the root of the issuer
+String orgUrl = new URL(new URL(issuer), "/").toString();
+
+ModelAndView mav = new ModelAndView("login");
+mav.addObject(STATE, state);
+mav.addObject(NONCE, nonce);
+mav.addObject(SCOPES, oktaOAuth2Properties.getScopes());
+mav.addObject(OKTA_BASE_URL, orgUrl);
+mav.addObject(OKTA_CLIENT_ID, oktaOAuth2Properties.getClientId());
+mav.addObject(INTERACTION_HANDLE, idxClientContext.getInteractionHandle());
+mav.addObject(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+mav.addObject(CODE_CHALLENGE, idxClientContext.getCodeChallenge());
+mav.addObject(CODE_CHALLENGE_METHOD, CODE_CHALLENGE_METHOD_VALUE);
+
+// from ClientRegistration.redirectUriTemplate, if the template is change you must update this
+mav.addObject(REDIRECT_URI,
+   request.getScheme() + "://" + request.getServerName() + ":" + request.getServerPort() +
+   request.getContextPath() + "/authorization-code/callback"
+);
+mav.addObject(ISSUER_URI, issuer);
+
+session.setAttribute(CODE_VERIFIER, idxClientContext.getCodeVerifier());
+```
+
+### Step 6: Request tokens from Okta
+
+Use the interaction code and code verifier to request for tokens. See helper class [exchangeCodeForToken()](https://github.com/okta/okta-idx-java/blob/f9378d48d39c10c76294e079f35214bbef3a02cd/samples/embedded-sign-in-widget/src/main/java/com/okta/spring/example/HelperUtil.java#L80).
+
+```java
+final JsonNode jsonNode = helperUtil.exchangeCodeForToken(interactionCode, codeVerifier);
+final OAuth2AccessToken oAuth2AccessToken = helperUtil.buildOAuth2AccessToken(jsonNode);
+final OAuth2RefreshToken oAuth2RefreshToken = helperUtil.buildOAuth2RefreshToken(jsonNode);
+```
+
+The Spring security framework doesn't understand Okta’s interaction code flow. Therefore, the app needs to intercept Spring’s OAuth authentication code flow, exchange the interaction code obtained from Okta for an access token, fill the user profile attributes, and construct [OAuth2AuthenticationToken.java](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/OAuth2AuthenticationToken.java) to hand over authentication flow back to Spring.
+
+### Step 7: Retrieve user profile
+
+Retrieve the user profile attributes with the access token object and populate Spring framework’s [OAuth2AuthenticationToken](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/OAuth2AuthenticationToken.java)  object reference for Spring to continue with the rest of the authentication flow.
```

Comment:
Would be nice to link HelperUtil class's `getUserAttributes` method here.

Reply:
```suggestion
Retrieve the user profile attributes with the access token object and populate Spring framework’s [OAuth2AuthenticationToken](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/authentication/OAuth2AuthenticationToken.java)  object reference for Spring to continue with the rest of the authentication flow. See helper class method [getUserAttributes()](https://github.com/okta/okta-idx-java/blob/f9378d48d39c10c76294e079f35214bbef3a02cd/samples/embedded-sign-in-widget/src/main/java/com/okta/spring/example/HelperUtil.java#L67) for details. 
```